File: services/recommendation-service/src/recommendation/recommendation.service.ts
import { randomUUID } from 'node:crypto';

import {
  Injectable,
  Logger,
  NotFoundException,
} from '@nestjs/common';
import { ConfigService } from '@nestjs/config';

import { DatabaseService } from '../database/database.service';
import { MetricsService } from '../metrics/metrics.service';
import { RequestContextService } from '../observability/request-context.service';
import {
  FeedbackRow,
  PreferenceVectorRow,
  RecommendationItem,
  RecommendationRow,
} from './entities/recommendation.entity';
import { GenerateRecommendationDto } from './dto/generate-recommendation.dto';
import { FeedbackDto } from './dto/feedback.dto';
import { TrendingQueryDto } from './dto/trending-query.dto';
import { IndexDto } from './dto/index.dto';

interface ScoredCandidate {
  id: string;
  type: string;
  title: string;
  summary: string | null;
  source: string;
  metadata: Record<string, unknown>;
  distance: number;
  importance: number;
  recency: number;
  usageCount: number;
}

@Injectable()
export class RecommendationService {
  private readonly logger = new Logger(RecommendationService.name);

  constructor(
    private readonly db: DatabaseService,
    private readonly config: ConfigService,
    private readonly metrics: MetricsService,
    private readonly requestContext: RequestContextService,
  ) {}

  async generate(dto: GenerateRecommendationDto): Promise<RecommendationRow> {
    const tenantId = this.requestContext.tenantId ?? dto.tenantId ?? 'default';
    const topN = dto.topN ?? 10;

    const [memoryEmbedding, nodeEmbedding] = await Promise.all([
      this.getUserPreferenceVector(dto.userId, tenantId, 'memory'),
      this.getUserPreferenceVector(dto.userId, tenantId, 'node'),
    ]);

    const candidates = await this.fetchCandidates(tenantId, memoryEmbedding, nodeEmbedding, topN * 3);

    let ranked = this.rank(candidates, dto.contextTags ?? []);

    if (dto.personalize) {
      ranked = await this.personalize(dto.userId, tenantId, ranked, topN);
    }

    const items: RecommendationItem[] = ranked.slice(0, topN).map((c) => ({
      id: c.id,
      type: c.type,
      title: c.title,
      summary: c.summary ?? undefined,
      score: this.compositeScore(c),
      source: c.source,
      metadata: c.metadata,
    }));

    const id = randomUUID();
    const result = await this.db.query<RecommendationRow>(
      `INSERT INTO recommendations (id, user_id, tenant_id, items, model_version)
       VALUES ($1, $2, $3, $4::jsonb, $5)
       RETURNING *`,
      [
        id,
        dto.userId,
        tenantId,
        JSON.stringify(items),
        dto.personalize ? 'ai-personalized-v1' : 'vector-v1',
      ],
    );

    this.metrics.recommendationsGenerated.inc({ tenant_id: tenantId });
    this.logger.log({ userId: dto.userId, count: items.length }, 'Recommendations generated');
    return result.rows[0]!;
  }

  async getByUser(userId: string): Promise<RecommendationRow[]> {
    const tenantId = this.requestContext.tenantId ?? 'default';
    const result = await this.db.query<RecommendationRow>(
      `SELECT * FROM recommendations WHERE user_id = $1 AND tenant_id = $2 ORDER BY created_at DESC`,
      [userId, tenantId],
    );
    return result.rows;
  }

  async recordFeedback(dto: FeedbackDto): Promise<FeedbackRow> {
    const tenantId = this.requestContext.tenantId ?? 'default';
    const userId = this.requestContext.userId ?? 'anonymous';

    const exists = await this.db.query<{ id: string }>(
      `SELECT id FROM recommendations WHERE id = $1`,
      [dto.recommendationId],
    );
    if (exists.rows.length === 0) {
      throw new NotFoundException(`Recommendation "${dto.recommendationId}" not found`);
    }

    const id = randomUUID();
    const result = await this.db.query<FeedbackRow>(
      `INSERT INTO recommendation_feedback (id, recommendation_id, user_id, tenant_id, useful)
       VALUES ($1, $2, $3, $4, $5)
       RETURNING *`,
      [id, dto.recommendationId, userId, tenantId, dto.useful],
    );

    this.metrics.feedbackRecorded.inc({ tenant_id: tenantId, useful: String(dto.useful) });
    return result.rows[0]!;
  }

  async getTrending(queryDto: TrendingQueryDto): Promise<RecommendationItem[]> {
    const tenantId = this.requestContext.tenantId ?? queryDto.tenantId ?? 'default';
    const limit = queryDto.limit ?? 20;

    const memories = await this.db.query<RecommendationItem & { usage_count: string }>(
      `SELECT id, 'memory' AS type, title, summary, metadata,
              COALESCE((metadata->>'usageCount')::int, 0) AS usage_count
       FROM memories
       WHERE tenant_id = $1 AND status != 'DELETED'
       ORDER BY usage_count DESC
       LIMIT $2`,
      [tenantId, limit],
    );

    const nodes = await this.db.query<RecommendationItem & { usage_count: string }>(
      `SELECT id, 'knowledge_node' AS type, label AS title, description AS summary, properties AS metadata,
              COALESCE((properties->>'usageCount')::int, 0) AS usage_count
       FROM knowledge_nodes
       WHERE tenant_id = $1
       ORDER BY usage_count DESC
       LIMIT $2`,
      [tenantId, limit],
    );

    const merged = [...memories.rows, ...nodes.rows]
      .map((r) => ({
        id: r.id,
        type: r.type,
        title: r.title,
        summary: r.summary ?? undefined,
        score: Number(r.usage_count),
        source: r.type,
        metadata: r.metadata,
      }))
      .sort((a, b) => b.score - a.score)
      .slice(0, limit);

    return merged;
  }

  async reindex(dto: IndexDto): Promise<{ userIds: number; updatedAt: string }> {
    const tenantId = this.requestContext.tenantId ?? dto.tenantId ?? 'default';

    const where = dto.userId ? 'WHERE user_id = $1 AND tenant_id = $2' : 'WHERE tenant_id = $1';
    const params = dto.userId ? [dto.userId, tenantId] : [tenantId];

    const users = await this.db.query<{ user_id: string }>(
      `SELECT DISTINCT user_id FROM memories ${where}`,
      params,
    );

    for (const row of users.rows) {
      const vector = await this.computeAggregatePreference(row.user_id, tenantId);
      await this.db.query<PreferenceVectorRow>(
        `INSERT INTO user_preference_vectors (user_id, tenant_id, preference_vector, updated_at)
         VALUES ($1, $2, $3::vector, NOW())
         ON CONFLICT (user_id, tenant_id)
         DO UPDATE SET preference_vector = $3::vector, updated_at = NOW()`,
        [row.user_id, tenantId, `[${vector.join(',')}]`],
      );
    }

    this.metrics.indexRebuilds.inc({ tenant_id: tenantId });
    this.logger.log({ tenantId, count: users.rows.length }, 'Preference index rebuilt');
    return { userIds: users.rows.length, updatedAt: new Date().toISOString() };
  }

  private async getUserPreferenceVector(
    userId: string,
    tenantId: string,
    _source: 'memory' | 'node',
  ): Promise<number[]> {
    const result = await this.db.query<{ preference_vector: string }>(
      `SELECT preference_vector FROM user_preference_vectors WHERE user_id = $1 AND tenant_id = $2`,
      [userId, tenantId],
    );

    const dim = this.config.get<number>('ai.embeddingDimension') ?? 1536;

    if (result.rows.length === 0 || !result.rows[0]!.preference_vector) {
      return new Array(dim).fill(0);
    }

    return this.parseVector(result.rows[0]!.preference_vector);
  }

  private async fetchCandidates(
    tenantId: string,
    memoryEmbedding: number[],
    nodeEmbedding: number[],
    limit: number,
  ): Promise<ScoredCandidate[]> {
    const memoryVec = `[${memoryEmbedding.join(',')}]`;
    const nodeVec = `[${nodeEmbedding.join(',')}]`;

    const memoryResult = await this.db.query<
      ScoredCandidate & { embedding: unknown }
    >(
      `SELECT id, 'memory' AS type, title, summary, metadata,
              COALESCE((metadata->>'importance')::float, 0.5) AS importance,
              EXTRACT(EPOCH FROM NOW() - created_at) / 86400.0 AS recency,
              COALESCE((metadata->>'usageCount')::int, 0) AS usage_count,
              1 - (embedding <=> $1::vector) AS distance
       FROM memories
       WHERE tenant_id = $2 AND status != 'DELETED' AND embedding IS NOT NULL
       ORDER BY distance DESC
       LIMIT $3`,
      [memoryVec, tenantId, limit],
    );

    const nodeResult = await this.db.query<
      ScoredCandidate & { embedding: unknown }
    >(
      `SELECT id, 'knowledge_node' AS type, label AS title, description AS summary, properties AS metadata,
              COALESCE((properties->>'importance')::float, 0.5) AS importance,
              EXTRACT(EPOCH FROM NOW() - created_at) / 86400.0 AS recency,
              COALESCE((properties->>'usageCount')::int, 0) AS usage_count,
              1 - (embedding <=> $1::vector) AS distance
       FROM knowledge_nodes
       WHERE tenant_id = $2 AND embedding IS NOT NULL
       ORDER BY distance DESC
       LIMIT $3`,
      [nodeVec, tenantId, limit],
    );

    const toCandidate = (r: ScoredCandidate): ScoredCandidate => ({
      id: r.id,
      type: r.type,
      title: r.title,
      summary: r.summary,
      source: r.type,
      metadata: typeof r.metadata === 'string' ? JSON.parse(r.metadata) : r.metadata,
      distance: r.distance,
      importance: r.importance,
      recency: r.recency,
      usageCount: r.usageCount,
    });

    return [...memoryResult.rows, ...nodeResult.rows].map(toCandidate);
  }

  private rank(candidates: ScoredCandidate[], contextTags: string[]): ScoredCandidate[] {
    const weightDistance = 0.5;
    const weightImportance = 0.3;
    const weightRecency = 0.1;
    const weightUsage = 0.1;
    const maxUsage = Math.max(1, ...candidates.map((c) => c.usageCount));
    const maxRecency = Math.max(1, ...candidates.map((c) => c.recency));

    return candidates
      .map((c) => {
        const recencyScore = 1 / (1 + c.recency / maxRecency);
        const usageScore = c.usageCount / maxUsage;
        const ctxBoost = this.contextBoost(c, contextTags);
        const score =
          (weightDistance * c.distance +
            weightImportance * c.importance +
            weightRecency * recencyScore +
            weightUsage * usageScore) *
          (1 + ctxBoost);
        return { ...c, distance: score };
      })
      .sort((a, b) => b.distance - a.distance);
  }

  private contextBoost(c: ScoredCandidate, contextTags: string[]): number {
    if (contextTags.length === 0) return 0;
    const tags: string[] = Array.isArray(c.metadata?.tags)
      ? (c.metadata!.tags as string[])
      : [];
    const overlap = tags.filter((t) => contextTags.includes(t)).length;
    return overlap / contextTags.length;
  }

  private compositeScore(c: ScoredCandidate): number {
    return Number(c.distance.toFixed(4));
  }

  private async personalize(
    userId: string,
    tenantId: string,
    ranked: ScoredCandidate[],
    topN: number,
  ): Promise<ScoredCandidate[]> {
    const url = this.config.get<string>('ai.serviceUrl') as string;
    const endpoint = this.config.get<string>('ai.personalizationEndpoint') as string;

    const start = Date.now();
    try {
      const response = await fetch(`${url}${endpoint}`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          userId,
          tenantId,
          candidates: ranked.slice(0, topN * 2).map((c) => ({
            id: c.id,
            type: c.type,
            title: c.title,
          })),
        }),
      });
      if (!response.ok) {
        throw new Error(`AI personalization returned ${response.status}`);
      }
      const body = (await response.json()) as { ranking?: string[] };
      const order = body.ranking ?? [];
      const byId = new Map(ranked.map((c) => [c.id, c]));
      const reordered = order
        .map((id) => byId.get(id))
        .filter((c): c is ScoredCandidate => Boolean(c));
      const remainder = ranked.filter((c) => !order.includes(c.id));
      this.metrics.modelLatency.observe({ tenant_id: tenantId }, (Date.now() - start) / 1000);
      return [...reordered, ...remainder];
    } catch (err) {
      this.logger.warn({ err: (err as Error).message }, 'Personalization failed, using vector ranking');
      return ranked;
    }
  }

  private async computeAggregatePreference(userId: string, tenantId: string): Promise<number[]> {
    const dim = this.config.get<number>('ai.embeddingDimension') ?? 1536;
    const result = await this.db.query<{ avg: string }>(
      `SELECT AVG(embedding)::text AS avg
       FROM memories
       WHERE tenant_id = $1 AND user_id = $2 AND status != 'DELETED' AND embedding IS NOT NULL`,
      [tenantId, userId],
    );

    const avg = result.rows[0]?.avg;
    if (!avg) {
      return new Array(dim).fill(0);
    }
    return this.parseVector(avg);
  }

  private parseVector(raw: string): number[] {
    const trimmed = raw.replace(/^\[/, '').replace(/\]$/, '').trim();
    if (!trimmed) return [];
    return trimmed.split(',').map((s) => Number(s.trim()));
  }
}

File: apps/api/src/api/services/recommendation_service.py
import uuid
from typing import Any

from sqlalchemy import text

from ..services.llm_service import llm_service


class RecommendationService:
    async def _compute_embedding(self, text_content: str) -> list[float]:
        try:
            return await llm_service.generate_embedding(text_content)
        except Exception:
            pass
        return [0.0] * 1536

    async def generate(self, dto, db):
        user_id = dto.user_id
        tenant_id = dto.tenant_id or "default"
        top_n = dto.top_n
        context_tags = dto.context_tags or []

        pref_vector_result = await db.execute(
            text("""
                SELECT preference_vector FROM user_preference_vectors
                WHERE user_id = :user_id AND tenant_id = :tenant_id
            """),
            {"user_id": user_id, "tenant_id": tenant_id},
        )
        pref_row = pref_vector_result.fetchone()

        embedding = [0.0] * 1536
        if pref_row and pref_row[0]:
            raw = pref_row[0]
            if isinstance(raw, str):
                raw = raw.strip("[]").split(",")
            embedding = [float(v) for v in raw]

        embedding_str = "[" + ",".join(f"{v}" for v in embedding) + "]"

        memories_result = await db.execute(
            text("""
                SELECT id::text, 'memory' AS type, title, summary, metadata,
                       COALESCE((metadata->>'importance')::float, 0.5) AS importance,
                       EXTRACT(EPOCH FROM NOW() - created_at) / 86400.0 AS recency,
                       COALESCE((metadata->>'usageCount')::int, 0) AS usage_count,
                       1 - (embedding <=> :embedding::vector) AS distance
                FROM memories
                WHERE tenant_id = :tenant_id AND status != 'deleted' AND embedding IS NOT NULL
                ORDER BY distance DESC
                LIMIT :top_n
            """),
            {"embedding": embedding_str, "tenant_id": tenant_id, "top_n": top_n * 3},
        )
        memory_rows = memories_result.fetchall() or []

        nodes_result = await db.execute(
            text("""
                SELECT id::text, type, label AS title, description AS summary, properties AS metadata,
                       importance,
                       EXTRACT(EPOCH FROM NOW() - created_at) / 86400.0 AS recency,
                       COALESCE((properties->>'usageCount')::int, 0) AS usage_count,
                       1 - (embedding <=> :embedding::vector) AS distance
                FROM knowledge_nodes
                WHERE tenant_id = :tenant_id AND embedding IS NOT NULL
                ORDER BY distance DESC
                LIMIT :top_n
            """),
            {"embedding": embedding_str, "tenant_id": tenant_id, "top_n": top_n * 3},
        )
        node_rows = nodes_result.fetchall() or []

        candidates = []
        for r in memory_rows:
            candidates.append(self._build_item(r, "memory"))

        for r in node_rows:
            candidates.append(self._build_item(r, "knowledge_node"))

        if context_tags:
            for c in candidates:
                meta = c.get("metadata") or {}
                tags = meta.get("tags") or []
                tag_overlap = len(set(tags) & set(context_tags))
                c["context_boost"] = min(tag_overlap * 0.1, 0.3)
            for c in candidates:
                c["score"] = (
                    c.get("distance", 0) * 0.5
                    + c.get("importance", 0.5) * 0.3
                    + c.get("recency_score", 0) * 0.1
                    + c.get("usage_score", 0) * 0.1
                    + c.get("context_boost", 0)
                )
        else:
            for c in candidates:
                c["score"] = (
                    c.get("distance", 0) * 0.5
                    + c.get("importance", 0.5) * 0.3
                    + c.get("recency_score", 0) * 0.1
                    + c.get("usage_score", 0) * 0.1
                )

        candidates.sort(key=lambda x: x["score"], reverse=True)

        top_items = candidates[:top_n]

        items_json = [
            {
                "id": c["id"],
                "type": c["type"],
                "title": c["title"],
                "summary": c.get("summary"),
                "score": round(c["score"], 6),
                "source": c.get("source", "vector"),
                "metadata": c.get("metadata"),
            }
            for c in top_items
        ]

        rec_id = uuid.uuid4()
        model_version = "v1"

        result = await db.execute(
            text("""
                INSERT INTO recommendations (id, user_id, tenant_id, items, model_version)
                VALUES (:id, :user_id, :tenant_id, :items::jsonb, :model_version)
                RETURNING id, user_id, tenant_id, items::text, model_version, created_at
            """),
            {
                "id": rec_id,
                "user_id": user_id,
                "tenant_id": tenant_id,
                "items": items_json,
                "model_version": model_version,
            },
        )
        return result.fetchone()

    def _build_item(self, row, source: str) -> dict:
        distance = float(row.distance) if row.distance is not None else 0
        importance = float(row.importance) if row.importance is not None else 0.5
        recency_days = float(row.recency) if row.recency is not None else 365
        usage_count = int(row.usage_count) if row.usage_count is not None else 0

        recency_score = max(0, 1.0 - recency_days / 365.0)
        usage_score = min(usage_count / 100.0, 1.0)

        return {
            "id": str(row.id),
            "type": str(row.type),
            "title": str(row.title),
            "summary": str(row.summary) if row.summary else None,
            "metadata": dict(row.metadata) if row.metadata else {},
            "distance": distance,
            "importance": importance,
            "recency_score": recency_score,
            "usage_score": usage_score,
            "source": source,
        }

    async def get_by_user(self, user_id: str, db):
        result = await db.execute(
            text("""
                SELECT id, user_id, tenant_id, items::text, model_version, created_at
                FROM recommendations
                WHERE user_id = :user_id
                ORDER BY created_at DESC
            """),
            {"user_id": user_id},
        )
        return result.fetchall()

    async def record_feedback(self, dto, db):
        rec_result = await db.execute(
            text("SELECT id FROM recommendations WHERE id = :rid"),
            {"rid": uuid.UUID(dto.recommendation_id)},
        )
        if not rec_result.fetchone():
            return None

        feedback_id = uuid.uuid4()
        row_result = await db.execute(
            text("""
                INSERT INTO recommendation_feedback (id, recommendation_id, useful)
                VALUES (:id, :recommendation_id, :useful)
                RETURNING id, recommendation_id, useful, created_at
            """),
            {
                "id": feedback_id,
                "recommendation_id": uuid.UUID(dto.recommendation_id),
                "useful": dto.useful,
            },
        )
        return row_result.fetchone()

    async def get_trending(self, limit: int, tenant_id: str | None, db):
        params: dict[str, Any] = {"limit": limit}

        tenant_clause = ""
        if tenant_id:
            tenant_clause = "AND tenant_id = :tenant_id"
            params["tenant_id"] = tenant_id

        memories_result = await db.execute(
            text(f"""
                SELECT id::text AS item_id, 'memory' AS item_type, title,
                       summary, metadata, COALESCE((metadata->>'usageCount')::int, 0) AS usage_count
                FROM memories
                WHERE status != 'deleted' {tenant_clause}
                ORDER BY usage_count DESC
                LIMIT :limit
            """),
            params,
        )
        memory_rows = memories_result.fetchall() or []

        nodes_result = await db.execute(
            text(f"""
                SELECT id::text AS item_id, 'knowledge_node' AS item_type, label AS title,
                       description AS summary, properties AS metadata,
                       COALESCE((properties->>'usageCount')::int, 0) AS usage_count
                FROM knowledge_nodes
                WHERE 1=1 {tenant_clause}
                ORDER BY usage_count DESC
                LIMIT :limit
            """),
            params,
        )
        node_rows = nodes_result.fetchall() or []

        merged = []
        for r in memory_rows:
            merged.append({
                "id": str(r.item_id),
                "type": str(r.item_type),
                "title": str(r.title),
                "summary": str(r.summary) if r.summary else None,
                "metadata": dict(r.metadata) if r.metadata else {},
                "score": float(r.usage_count),
                "source": "memory",
            })
        for r in node_rows:
            merged.append({
                "id": str(r.item_id),
                "type": str(r.item_type),
                "title": str(r.title),
                "summary": str(r.summary) if r.summary else None,
                "metadata": dict(r.metadata) if r.metadata else {},
                "score": float(r.usage_count),
                "source": "knowledge_node",
            })

        merged.sort(key=lambda x: x["score"], reverse=True)
        return merged[:limit]

    async def reindex(self, user_id: str | None, tenant_id: str | None, db):
        params: dict[str, Any] = {}

        user_clause = ""
        if user_id:
            user_clause = "AND m.user_id = :user_id"
            params["user_id"] = user_id

        tenant_clause = ""
        if tenant_id:
            tenant_clause = "AND m.tenant_id = :tenant_id"
            params["tenant_id"] = tenant_id

        users_result = await db.execute(
            text(f"""
                SELECT DISTINCT m.user_id
                FROM memories m
                WHERE m.user_id IS NOT NULL AND m.embedding IS NOT NULL
                {user_clause} {tenant_clause}
            """),
            params,
        )
        user_rows = users_result.fetchall() or []

        results = []
        for row in user_rows:
            uid = row[0]
            avg_result = await db.execute(
                text("""
                    SELECT AVG(embedding)::text FROM memories
                    WHERE user_id = :uid AND tenant_id = :tid AND embedding IS NOT NULL
                """),
                {"uid": uid, "tid": tenant_id or "default"},
            )
            avg_row = avg_result.fetchone()
            if avg_row and avg_row[0]:
                await db.execute(
                    text("""
                        INSERT INTO user_preference_vectors (user_id, tenant_id, preference_vector)
                        VALUES (:uid, :tid, :vector::vector)
                        ON CONFLICT (user_id, tenant_id)
                        DO UPDATE SET preference_vector = :vector::vector, updated_at = NOW()
                    """),
                    {"uid": uid, "tid": tenant_id or "default", "vector": avg_row[0]},
                )
                results.append({"user_id": str(uid), "status": "reindexed"})

        return results


recommendation_service = RecommendationService()
